### PR TITLE
Validate Transitions

### LINKED ISSUE #561: Validate transitions
A production wallet needs to guarantee that it can launch a force-move/refutation/conclusion. This requires a valid round (or at least a valid pair) of commitments. That is, each successive pair of commitments forms a valid transition.

Currently the wallet does not guarantee this. The primary reason is that a reducer has to be a pure, sync function. Thus, it can't call a smart contract function to validate transitions according to the "ground-truth" rules.

### PR DESCRIPTION
**Pull request by lalexgap**
Calls `validateTransition` from the `message-listener` whenever receiving a `commitment`.

Fixes #561.

### CODE REVIEW

File: packages/wallet/src/redux/sagas/message-listener.ts [hunk 1]
```diff
@@ -94,6 +108,59 @@ export function* messageListener() {
...
+  // If we're receiving the first commitment there's nothing stored to validate against
+  if (signedCommitments[0].commitment.turnNum > 0) {
+    const latestStoredCommitment: Commitment = yield select(
+      selectors.getLastCommitmentForChannel,
+      channelId,
+    );
+    const newCommitments = signedCommitments.filter(
+      signedCommitment => signedCommitment.commitment.turnNum > latestStoredCommitment.turnNum,
+    );
+    if (newCommitments.length > 0) {
```
[the remark this answers is not in the document]
Comment (andrewgordstewart) [reply packages/wallet/src/redux/sagas/message-listener.ts]: Considering that, we can just do
```
signedCommitments.filter(
  signedCommitment => signedCommitment.commitment.turnNum > latestStoredCommitment.turnNum,
).reduce( [above code] )
```

File: packages/wallet/src/redux/sagas/message-listener.ts [hunk 2]
```diff
@@ -94,6 +108,59 @@ export function* messageListener() {
   }
 }
 
+function* validateTransitionForCommitments(signedCommitments: SignedCommitment[]) {
+  const channelId = getChannelId(signedCommitments[0].commitment);
+  // If we're receiving the first commitment there's nothing stored to validate against
+  if (signedCommitments[0].commitment.turnNum > 0) {
```
Comment (andrewgordstewart) [inline packages/wallet/src/redux/sagas/message-listener.ts]: I think we would want to ask directly if we have something stored to validate against. If B sends [{turnNum: 0}, {turnNum: 1}]` to A, then we wouldn't enter this block.

File: packages/wallet/src/redux/sagas/message-listener.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (lalexgap) [reply packages/wallet/src/redux/sagas/message-listener.ts]: `yield` is not allowed inside a reducer function.

File: packages/wallet/src/redux/sagas/message-listener.ts (hunk 2 above)
Comment (lalexgap) [reply packages/wallet/src/redux/sagas/message-listener.ts]: Updated

File: packages/wallet/src/redux/sagas/message-listener.ts [hunk 3]
```diff
@@ -94,6 +108,60 @@ export function* messageListener() {
   }
 }
 
+function* validateTransitionForCommitments(signedCommitments: SignedCommitment[]) {
+  const channelId = getChannelId(signedCommitments[0].commitment);
+
+  const storedCommitmentExists = yield select(selectors.doesACommitmentExistForChannel, channelId);
```
Comment (andrewgordstewart) [inline packages/wallet/src/redux/sagas/message-listener.ts]: An alternative here is to rewrite `getLastCommitmentForChannel` to throw when the channel isn't stored, catch that error, and do nothing.

(Or have it return `undefined`, and if it's undefined, do nothing)

Comment (lalexgap) [reply packages/wallet/src/redux/sagas/message-listener.ts]: I think I'd prefer the extra `selector` so I'm going to keep it that way for now.

**Pull request merged**